source-kafka/src/connector.rs

```diff
@@ -39,3 +44,95 @@ pub enum Command {
         state: Option<PathBuf>,
     },
 }
+
+impl Command {
+    pub fn execute<C: Connector>(&self) -> color_eyre::Result<()> {
+        let stdout_guard = stdout();
+        let mut stdout = Box::new(stdout_guard.lock());
+
+        let result = match self {
+            Command::Spec => C::spec(&mut stdout),
+            Command::Check { config } => {
+                C::check(&mut stdout, C::Config::parse(open_file(config)?)?)
+            }
+            Command::Discover { config } => {
+                C::discover(&mut stdout, C::Config::parse(open_file(config)?)?)
+            }
+            Command::Read {
+                config,
+                catalog,
+                state,
+            } => {
+                let state = match state {
+                    None => None,
+                    Some(s) => {
+                        // Can't use Option::map with `?`, so we get to match here.
+                        Some(C::State::parse(open_file(s)?)?)
+                    }
+                };
+
+                C::read(
+                    &mut stdout,
+                    C::Config::parse(open_file(config)?)?,
+                    C::ConfiguredCatalog::parse(open_file(catalog)?)?,
+                    state,
+                )
+            }
+        };
+
+        match result {
+            Err(e) if e.is::<DocumentStreamError>() => {
+                // Stdout has been closed, so we should gracefully shut down rather than panicking.
+                Ok(())
+            }
+            otherwise => otherwise,
+        }
+    }
+}
+
+fn open_file(path: impl AsRef<Path>) -> color_eyre::Result<File> {
+    File::open(&path).wrap_err(format!("failed to read {:?}", path.as_ref()))
+}
+
+pub trait ConnectorConfig: Sized {
+    type Error: std::error::Error + Send + Sync + 'static;
+
+    fn parse(reader: impl Read) -> Result<Self, Self::Error>;
+}
+
+pub trait Connector {
+    type Config: ConnectorConfig;
+    type ConfiguredCatalog: ConnectorConfig;
+    type State: ConnectorConfig;
+
+    fn spec(output: &mut dyn Write) -> color_eyre::Result<()>;
+    fn check(output: &mut dyn Write, config: Self::Config) -> color_eyre::Result<()>;
+    fn discover(output: &mut dyn Write, config: Self::Config) -> color_eyre::Result<()>;
+    fn read(
+        output: &mut dyn Write,
+        config: Self::Config,
+        catalog: Self::ConfiguredCatalog,
+        state: Option<Self::State>,
+    ) -> color_eyre::Result<()>;
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum DocumentStreamError {
```

Comment:
I think maybe `StdoutError` might be a better name? As it is, it's unclear whether this is an error in reading or writing, and it was a bit confusing to read as part of `execute` above.